### ISSUE TITLE
Execution hangs when browser window goes to background

### ISSUE THREAD

**Issue by printercu** [puppeteer/puppeteer#3156]
### Steps to reproduce

* Puppeteer version: 1.6.2
* Chromium: 69.0.3494.0 (Developer Build)
* Platform / OS version: macos sierra
* Node.js version: v10.8.0

**What steps will reproduce the problem?**

* Run the following script. It opens google and submits form, repeating 10 times.

```js
const puppeteer = require('puppeteer')

main()

async function main() {
  const browser = await puppeteer.launch({
    args: [
      // all related chromium flags I could find
      '--disable-background-timer-throttling',
      '--disable-renderer-backgrounding',
      '--override-plugin-power-saver-for-testing=never',
      '--disable-extensions-http-throttling',
    ],
    headless: false
  })
  const timeout = 5000
  const page = await browser.newPage()
  for (let i = 0; i < 10; i++) {
    console.log(`navigating ${i}`)
    await page.goto('https://google.com', {timeout})
    const input = await page.$('input[type="text"]')
    await input.type('test')
    await Promise.all([
      page.waitForNavigation({timeout}),
      page.evaluate(() => document.querySelector('input[type="submit"]').click())
    ])
  }
}
```

* Switch to any other fullscreen-size window, so puppeteer browser window is not visible.

**What is the expected result?**

Scripts proceeds till end without errors.

**What happens instead?**

UnhandledPromiseRejectionWarning: Error: Navigation Timeout Exceeded: 5000ms exceeded

**Some explorations**

This happens only when puppeteer window is not visible at all. When even couple of pixels of this window are visible everything works fine.

**Comment by vaibhav-kb** [comment]
**Your timeout isn't sufficient** 
Ex. Increase it to some large value like `const timeout = 30000` 
**OR** 
just remove `timeout` and it will work perfect without any errors

UnhandledPromiseRejectionWarning: Error: Navigation Timeout Exceeded: 5000ms exceeded => **means naviagtion isn't completing before 5000ms**

**Comment by printercu** [comment]
I have default timeout (30s) in production app where I've got this error initially. I've set lower value to not wait long when was trying to localize the problem. Though when I test now with `timeout: 30000` it freezes until I switch to puppeteer's browser and only then throws `Navigation Timeout Exceeded: 30000ms exceeded`. Same is when I remove timeout option.

**Comment by vaibhav-kb** [comment]
script (code snippet) you gave ran successfully when I added the `timeout = 30000`
but I realized one more thing that if your internet connection is too low, it will throw error even after `timeout = 30000`

**Comment by printercu** [comment]
This does NOT depend on internet connection. I run this script when the browser is ON TOP of other windows and it finishes fine and quick. Please make sure you understand that problem appears only when puppeteer browser is fully hidden under other windows. The better option is to switch to any other app with expanded window. If this does not happen to you anyway, please share your OS name and version.

**Comment by vaibhav-kb** [comment]
Ok...I will look again

**Comment by printercu** [comment]
With 30s timeout, make sure to wait a least 1min before switching back to puppeteer's window. Or just use lower timeout values. 5s is enough for me, because google opens in less than 2s.

**Comment by printercu** [comment]
I've just tested and found that this error appeared in 1.6.0 and still present in 1.7.0. It was ok in 1.5.0.

**Comment by vaibhav-kb** [comment]
I was just about to comment the same. It's working okay with less than
v1.5.0

**Comment by lehni** [comment]
Could this be linked to power-saving functions of Chrome? I noticed that when a little part of the browser window is visible, even when in the background, then execution is running. But as soon as the window is fully covered, execution seems to halt completely.

**Comment by aslushnikov** [comment]
@printercu The symptoms point to [AppNap](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/power_efficiency_guidelines_osx/AppNap.html) feature in Mac OS X. I'd try disabling it and see if it helps.

**Comment by lehni** [comment]
@aslushnikov yes I think that's exactly what I described above... I didn't know it was called `AppNap`. It can be disabled quite easily, see:
https://www.lifewire.com/control-how-mac-applications-use-app-nap-2260788

I will test this later today.

**Comment by lehni** [comment]
I haven't had time to test this yet, but I'm fairly certain #3148 is a duplicate of this issue

**Comment by lehni** [comment]
Ok, so I did check for the "App Nap" option, but it doesn't show on Chromium.

And in the article linked to above it says:

> If there's a Prevent App Nap checkbox present, you can put a check mark in the box to prevent naps, or remove a check mark to allow naps. If there's no checkbox, then the app isn't App Nap aware.

🤔 

<img width="377" alt="screenshot 2018-11-20 at 21 15 33" src="https://user-images.githubusercontent.com/213275/48800219-72202d00-ed09-11e8-9352-035a22a9e4f2.png">

**Comment by Fr0sT-Brutal** [comment]
The same issue on Windows and minimized Chrome window. Timeout extended to 20sec. First navigation completes and then things just hang until the window is shown.

**Comment by lehni** [comment]
Somewhat related, here's a workaround that I use on macOS in the meantime to not have my machine completely blocked by e2e tests running in Puppeteer: https://github.com/GoogleChrome/puppeteer/issues/1083#issuecomment-440422354

**Comment by Gdocal** [comment]
Can confirm the issue on Windows 10 with Chrome 72.0.3626.81. Mine freezes when attempt to click on element and chrome windows is in the background.  Is there are any workaround?

I tried to play with chrome://flags
![image](https://user-images.githubusercontent.com/11136265/52334401-80701200-29f7-11e9-8c28-2b1e841f5901.png)
no effect

**Comment by yoosif0** [comment]
Any solutions for that? This is just so weird.

**Comment by gsouf** [comment]
Confirm issue on linux too

**Comment by aslushnikov** [comment]
Ok so before we thought this is an APpNap behavior specific to Mac, but  I see there are now reports from Linux and WIndows as well.

Everybody: can you please share some script we can try running to see this happening?

**Comment by Gdocal** [comment]
> Ok so before we thought this is an APpNap behavior specific to Mac, but I see there are now reports from Linux and WIndows as well.
> 
> Everybody: can you please share some script we can try running to see this happening?

Not near pc right now . But I think any script which do some work will freeze when chrome windows is no visible . For me it always freezes on click and maybe some other operations

**Comment by printercu** [comment]
@aslushnikov I can not reproduce the issue on macos mojave with my snippet from https://github.com/GoogleChrome/puppeteer/issues/3156#issue-355496863.
I've tried both old 1.5 and new puppeteer.

**Comment by aslushnikov** [comment]
>  I can not reproduce the issue on macos mojave with my snippet from #3156 (comment).
I've tried both old 1.5 and new puppeteer.

Ok so if this strikes back - please feel free to create a new issue with a script that hangs for you when you minimize browser window. We'd be happy to look into that.

Closing this for now.

**Comment by Gdocal** [comment]
@aslushnikov @printercu Please reopen this annoying bug.
Here is code to reproduce it.
The issue is in IntersectionObserver which is used in _scrollIntoViewIfNeeded function which is used in Click.

Enviroment: 
OS: Windows 10
puppeteer: 1.19.0

```js
const puppeteer = require('puppeteer');

(async () => {
  var baseTime=Date.now();
  const browser = await puppeteer.launch({
    headless: false,
   //executablePath :"C:/Program Files (x86)/Google/Chrome/Application/chrome.exe"
  });
  const page = await browser.newPage();
  await page.goto('https://github.com/');
  page.on('console', msg => {
  for (let i = 0; i < msg.args().length; ++i)
    console.log( Date.now()-baseTime+' Console: '+` ${i}: ${msg.args()[i]}`);
});
 
  console.log('Please minimize chrome');
  process.stdout.write('\x07');
  await sleep(5000);
   var joinNode=await page.$('a[href *= join]');
    console.log(Date.now()-baseTime+' Before  _scrollIntoViewIfNeeded');
   await _scrollIntoViewIfNeeded(joinNode,true);
   console.log(Date.now()-baseTime+' after  _scrollIntoViewIfNeeded');
  await browser.close();
  
  function sleep(ms){
    return new Promise(resolve=>{
        setTimeout(resolve,ms)
    })
}
  async function  _scrollIntoViewIfNeeded(element,pageJavascriptEnabled) {
    const error = await  element.executionContext().evaluate(async(element, pageJavascriptEnabled) => {
      if (!element.isConnected)
        return 'Node is detached from document';
      if (element.nodeType !== Node.ELEMENT_NODE)
        return 'Node is not of type HTMLElement';
      // force-scroll if page's javascript is disabled.
      if (!pageJavascriptEnabled) {
        element.scrollIntoView({block: 'center', inline: 'center', behavior: 'instant'});
        return false;
      }
	  	console.log('before IntersectionObserver');
      const visibleRatio = await new Promise(resolve => {
        const observer = new IntersectionObserver(entries => {
			console.log('IntersectionObserver callback (ISSUE IS HERE!!!!!!)');
          resolve(entries[0].intersectionRatio);
          observer.disconnect();
        });
        observer.observe(element);
      });
	  
	  console.log('after IntersectionObserver');
      if (visibleRatio !== 1.0)
        element.scrollIntoView({block: 'center', inline: 'center', behavior: 'instant'});
      return false;
    }, element, true);
	  console.log('error: '+error);
    if (error)
      throw new Error(error);
  }
})();
```

After windows beep sound and message in console 'Please minimize chrome'- please minimize chrome window.

My log is 

>Please minimize chrome
> 7134 Before  _scrollIntoViewIfNeeded
> 7138 Console:  0: JSHandle:before IntersectionObserver
> 22261 Console:  0: JSHandle:IntersectionObserver callback (ISSUE IS HERE!!!!!!)
> 22265 Console:  0: JSHandle:after IntersectionObserver
> error: false
> 22269 after  _scrollIntoViewIfNeeded

I waited 15 seconds before restore chrome window and immediately after this I received message 'IntersectionObserver callback (ISSUE IS HERE!!!!!!)'

But this ossie occur not only when window is minimized, also when windows is in background

Please fix it. Because it is really anoying issue. 

Also I am afraid the root of this issue can cause some other issues which we even can't imagine now.
I am able to create workaround and just do not use  IntersectionObserver but how I can be sure that all other scripts in site will work as expected, and for example this behaviour will no be abused to detect puppeteer users.

**Comment by david-littlefield** [comment]
### [Solution]:

Add the following values to the array you assign to the args property in the options object. Now, you should be able to run Puppeteer while your window is in the foreground, background, or minimized. 

```
const chromeArgs = [
  '--disable-background-timer-throttling',
  '--disable-backgrounding-occluded-windows',
  '--disable-renderer-backgrounding'
];
```
Solution and explanation available in an article written by **Rohit Kumar**, @rkr090 at Big Binary:

[Debugging failing tests in puppeteer because of background tab](https://blog.bigbinary.com/2018/08/15/debugging-failing-tests-in-background-tab-in-puppeteer.html)
